### PR TITLE
Implicit sharing for QgsExpression

### PR DESCRIPTION
**Pull request by m-kuhn**
Title says it all.

So far one had to do

```
QgsExpression exp( "a>1" );
QgsExpression exp2( exp.expression() );
```

resulting in parsing the expression twice

Now one can do

```
QgsExpression exp( "a>1" );
QgsExpression exp2( exp );
```

The expression will only be parsed once.

Note that results of `prepare()` are not copied. The main effect of this is, that the column index cache is not copied.

The basic idea behind this is, that expressions should only always exist as a prepared copy when using them on one particular iterator, since that's the guaranteed lifetime of a `QgsFields` instance. And if an expression is `prepare()`'d it is detached.

```
// Will detach
exp.prepare( contextWithFields );
```

In short this means for the usage: keep a master copy of your expressions.
And whenever you use them with a new iterator, make a copy of them and prepare(). That's it.

### CODE REVIEW

**Comment by m-kuhn** [comment]
I would like to merge this soon to have it well tested before the next release. I'm quite confident that it is in a good shape.

If there are no objections raised, I'll merge this on Wednesday.

**Comment by NathanW2** [comment]
Merge it.

**Comment by m-kuhn** [comment]
Good idea ;) de1c319bf640ed45e57d4c93984
